src/com/carolinarollergirls/scoreboard/event/ScoreBoardEventProviderImpl.java

```diff
@@ -0,0 +1,446 @@
+            PermanentProperty sourceProperty, boolean readonly) {
+        ScoreBoardListener l = new ConditionalScoreBoardListener(sourceElement, sourceProperty,
+                new CopyScoreBoardListener(this, targetProperty));
+        sourceElement.addScoreBoardListener(l);
+        providers.put(l, sourceElement);
+        if (readonly) {
+            addWriteProtectionOverride(targetProperty, Flag.COPY);
+        } else {
+            reverseCopyListeners.put(targetProperty, 
+                    new CopyScoreBoardListener(sourceElement, sourceProperty, false));
+        }
+        set(targetProperty, sourceElement.get(sourceProperty), Flag.COPY);
+        return l;
+    }
+    protected ScoreBoardListener setCopy(final PermanentProperty targetProperty, ScoreBoardEventProvider indirectionElement,
+            PermanentProperty indirectionProperty, final PermanentProperty sourceProperty, boolean readonly) {
+        ScoreBoardListener l = new IndirectScoreBoardListener(indirectionElement, indirectionProperty, sourceProperty,
+                new CopyScoreBoardListener(this, targetProperty));
+        providers.put(l, null);
+        if (readonly) {
+            addWriteProtectionOverride(targetProperty, Flag.COPY);
+        } else {
+            ScoreBoardListener reverseListener = 
+            new ConditionalScoreBoardListener(indirectionElement, indirectionProperty,
+                    new ScoreBoardListener() {
+                        public void scoreBoardChange(ScoreBoardEvent event) {
+                            reverseCopyListeners.put(targetProperty,
+                                    new CopyScoreBoardListener((ScoreBoardEventProvider) event.getValue(),
+                                            sourceProperty, false));
+                        }
+                    });
+            indirectionElement.addScoreBoardListener(reverseListener);
+            reverseListener.scoreBoardChange(new ScoreBoardEvent(indirectionElement, indirectionProperty,
+                    indirectionElement.get(indirectionProperty), null));
+        }
+        return l;
+    }
+    protected RecalculateScoreBoardListener setRecalculated(PermanentProperty targetProperty) {
+        RecalculateScoreBoardListener l = new RecalculateScoreBoardListener(this, targetProperty);
+        providers.put(l, null);
+        return l;
+    }
+    protected InverseReferenceUpdateListener setInverseReference(Property localProperty, Property remoteProperty) {
+        InverseReferenceUpdateListener l = new InverseReferenceUpdateListener(this, localProperty, remoteProperty);
+        addScoreBoardListener(l);
+        return l;
+    }
+
+    public Object valueFromString(PermanentProperty prop, String sValue, Flag flag) {
+        synchronized (coreLock) {
+            @SuppressWarnings("rawtypes")
+            Class type = prop.getType();
+            if (type == TimeoutOwner.class) { return scoreBoard.getTimeoutOwner(sValue); }
```

Comment:
This is a layering violation, this should come from the object rather than this class

Reply:
I don't understand this comment.

Reply:
Check if the fromString method exists via an interface, rather than putting a list of classes here. We should only need to special case the POJOs.

Reply:
Specifically for TimeoutOwner this does not really work. As we don't have an Object yet, the fromString() method needs to be static. But the list of possible `TimeoutOwner`s includes `Team`s, which are only created at runtime in a specific `ScoreBoardImpl` object. So we would need to
a) inject the current `ScoreBoard` instance into a static variable in `TimeoutOwner` so `fromString()` can reference it.
b) use reflection to call the static `fromString()` method (this would also be required for the other classes)
I don't think that is an improvement over what we have now.

Reply:
You could pass the scoreBoard object as a parameter to valueFromString. You could also move this all over to the ScoreBoard object.

> I don't think that is an improvement over what we have now.

Two packages mutually depending on each other is to be avoided. The event provider code shouldn't know what types of object it's working with.

Reply:
If we want to hand the `scoreboard` object as a method parameter, it has to come from the `json` resp. `xml` package, two packages that also are referenced from `core` - we would again just shift the circular dependency around (and if we don't remove the other uses of the object in SBEPI, we don't even remove a dependency).

I argue that the `scoreboard` object is at the heart of this program and trying to isolate it in the `core` package is going to cause a huge amount of overhead and will significantly decrease the maintainability of the code, just for the sake of following a principle, as we can be rather certain that this class isn't going away in a refactoring. So this is a case, where a deviation from the principle is justified.

Reply:
The only thing even using this method is WS/xml, which can get it from the scoreboard object they already have directly. There's no reason for this code to be here.